rmw_opendds_cpp/src/rmw_publisher.cpp

```diff
@@ -110,195 +130,151 @@ rmw_create_publisher(
     return NULL;
   }
 
-  // TODO: TypeCode and susequent call to register external type have been removed
-  // May need to implement something similar after IDL type support is finalized
+  // TODO: Register TypeSupport here ????????????????????????
+  OpenDDSStaticSerializedDataTypeSupport_var ts = new OpenDDSStaticSerializedDataTypeSupportImpl();
+  if (ts->register_type(participant, "") != DDS::RETCODE_OK) {
+    RMW_SET_ERROR_MSG("failed to register OpenDDS type");
+    return NULL;
+  }
 
-  std::string type_name = _create_type_name(callbacks, "msg");
+  if (!topic_name || strlen(topic_name) == 0) {
+    RMW_SET_ERROR_MSG("publisher topic_name is null or empty");
+    return NULL;
+  }
 
-  void * info_buf = nullptr;
-  void * listener_buf = nullptr;
-  OpenDDSPublisherListener * publisher_listener = nullptr;
-  OpenDDSStaticPublisherInfo * publisher_info = nullptr;
-  rmw_publisher_t * publisher = nullptr;
-  std::string mangled_name = "";
-  OpenDDSStaticSerializedDataTypeSupport_var ts = new OpenDDSStaticSerializedDataTypeSupportImpl();
+  if (!qos_policies) {
+    RMW_SET_ERROR_MSG("qos_policies is null");
+    return NULL;
+  }
 
+  rmw_publisher_t * publisher = nullptr;
+  void * buf = nullptr;
   char * topic_str = nullptr;
-
   try {
-    try {
-      // Begin initializing elements
-      publisher = rmw_publisher_allocate();
-      if (!publisher) {
-        throw std::string("failed to allocate publisher");
-      }
+    // create rmw_publisher_t
+    publisher = rmw_publisher_allocate();
+    if (!publisher) {
+      throw std::string("failed to allocate publisher");
```

Comment:
While throw's aren't syntactically a goto, this code operates in much the same way. This isn't necessarily a critique. It's more of an observation.

Reply:
Yes, I thought about that issue when I reviewed it last week.  We may need to do another pass (after this PR) with a different error handling/cleanup approach.

Reply:
I don't see a clear action here. Since this PR is about getting the manual test to pass, I'll resolve this. We can revisit form (rather than function) at a later date.